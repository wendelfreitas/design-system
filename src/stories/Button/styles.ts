import styled, { css, DefaultTheme, keyframes } from 'styled-components';
import { ButtonProps } from '.';
import { Spinner8 as Spinner } from '@styled-icons/icomoon/Spinner8';

type Rename<T, K extends keyof T, N extends string> = Pick<
  T,
  Exclude<keyof T, K>
> &
  { [P in N]: T[K] };

type TypeProps = Rename<ButtonProps, 'type', 'buttonType'>;

type WrapperProps = { hasIcon: boolean } & Pick<
  TypeProps,
  'buttonType' | 'size' | 'fullWidth' | 'disabled'
>;

const wrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    height: 3rem;
    font-size: ${theme.font.sizes.xsmall};
  `,
  medium: (theme: DefaultTheme) => css`
    height: 4rem;
    font-size: ${theme.font.sizes.small};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,
  large: (theme: DefaultTheme) => css`
    height: 5rem;
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};
  `,
  primary: (theme: DefaultTheme) => css`
    background: ${theme.colors.primary};
    color: ${theme.colors.white};
    &:hover {
      background: ${theme.colors.lightPrimary};
    }
  `,

  disabled: (theme: DefaultTheme) => css`
    cursor: not-allowed;
    background: ${theme.colors.lightPrimary};

    &:hover {
      background: ${theme.colors.lightPrimary};
    }
  `,
  fullWidth: () =>
    css`
      width: 100%;
    `,
  withIcon: (theme: DefaultTheme) => css`
    svg {
      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `,
};

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, buttonType, size, fullWidth, hasIcon, disabled }) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    border: none;
    cursor: pointer;
    border-radius: ${theme.border.radius.normal};
    font-weight: 600;
    padding: ${theme.spacings.xxsmall};
    text-decoration: none;

    &:focus {
      outline: 0;
    }

    ${!!size && wrapperModifiers[size](theme)}
    ${!!buttonType && wrapperModifiers[buttonType](theme)}
    ${!!fullWidth && wrapperModifiers.fullWidth()}
    ${!!hasIcon && wrapperModifiers.withIcon(theme)}
    ${!!disabled && wrapperModifiers.disabled(theme)}
  `}
`;

const rotate = keyframes`
  from{
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
`;

export const Loading = styled(Spinner)`
  ${({ theme }) => css`
    color: ${theme.colors.white};
    margin-right: ${theme.spacings.xxsmall};
    animation: ${rotate} 0.5s linear infinite;
  `}
`;
