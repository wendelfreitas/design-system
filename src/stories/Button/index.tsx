import React from 'react';
import * as S from './styles';
import { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react';

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>;

export type ButtonProps = {
  /**
   * Can be set to primary or secondary
   */
  type?: 'primary';

  /**
   * How large should the button be?
   */
  size?: 'small' | 'medium' | 'large';
  /**
   * In the case of the button, have the maximum size of the container
   */
  fullWidth?: boolean;
  /**
   * An optional icon
   */
  icon?: React.ReactNode;
  /**
   * Optional click handler
   */
  onClick?: (event?: React.MouseEvent<HTMLButtonElement>) => void;
  as?: React.ElementType;
  htmlType?: string;
  /**
   * Disabled state of button
   */
  disabled?: boolean;
  /**
   * Set the loading status of button
   */
  loading?: boolean;
} & ButtonTypes;

/**
 * Primary UI component for user interaction
 */
const Button = ({
  children,
  icon,
  type = 'primary',
  size = 'medium',
  htmlType,
  disabled = false,
  loading = false,
  fullWidth = false,
  ...props
}: ButtonProps) => {
  return (
    <S.Wrapper
      size={size}
      fullWidth={fullWidth}
      hasIcon={!!icon}
      type={htmlType}
      buttonType={type}
      disabled={loading ? true : disabled}
      {...props}
    >
      {loading ? <S.Loading size={15} /> : icon}
      {!!children && <span>{children} FINAL TEST</span>}
    </S.Wrapper>
  );
};

export default Button;
