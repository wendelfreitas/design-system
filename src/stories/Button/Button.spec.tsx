import React from 'react';
import { screen } from '@testing-library/react';
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart';
import { renderWithTheme } from 'utils/tests/helpers';
import Button from '.';

describe('<Button />', () => {
  it('should render the medium size by default', () => {
    const { container } = renderWithTheme(<Button>Design System</Button>);
    expect(screen.getByRole('button', { name: /Design System/i })).toHaveStyle({
      height: '4rem',
      'font-size': '1.4rem',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the small size', () => {
    renderWithTheme(<Button size="small">Design System</Button>);
    expect(screen.getByRole('button', { name: /Design System/i })).toHaveStyle({
      height: '3rem',
      'font-size': '1.2rem',
    });
  });

  it('should render the large size', () => {
    renderWithTheme(<Button size="large">Design System</Button>);
    expect(screen.getByRole('button', { name: /Design System/i })).toHaveStyle({
      height: '5rem',
      padding: '0.8rem 4.8rem',
      'font-size': '1.6rem',
    });
  });

  it('should render a fullWidth version', () => {
    renderWithTheme(<Button fullWidth>Design System</Button>);
    expect(screen.getByRole('button', { name: /Design System/i })).toHaveStyle({
      width: '100%',
    });
  });

  it('should render an icon version', () => {
    renderWithTheme(
      <Button icon={<AddShoppingCart data-testid="icon" />}>
        Design System
      </Button>
    );

    expect(screen.getByText(/Design System/i)).toBeInTheDocument();
    expect(screen.getByTestId('icon')).toBeInTheDocument();
  });

  it('should render button as a link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Design System
      </Button>
    );

    expect(
      screen.getByRole('link', { name: /Design System/i })
    ).toHaveAttribute('href', '/link');
  });

  it('should render button with loading indicator', () => {
    renderWithTheme(<Button loading>Design System</Button>);

    expect(screen.getByText(/Design System/i)).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Design System/i })).toHaveStyle({
      'background-color': '#FF8084',
      cursor: 'not-allowed',
    });
  });
});
