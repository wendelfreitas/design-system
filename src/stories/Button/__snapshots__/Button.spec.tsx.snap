// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium size by default 1`] = `
<button
  class="sc-bdvvaa iVerbf"
>
  <span>
    Design System
     FINAL TEST
  </span>
</button>
`;
